File: stardroid-v1/app/src/main/java/com/google/android/stardroid/pushnav/PushNavTargetSender.java
/*
 * Copyright (c) 2026 Jasper Luetkens.
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 */
package com.google.android.stardroid.pushnav;

import android.util.Log;

import org.json.JSONObject;

import java.io.IOException;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/** Sends selected Sky Map targets to PushNav. */
public final class PushNavTargetSender {
  private static final String TAG = "PushNavTargetSender";
  private static final int CONNECT_TIMEOUT_MS = 3000;
  private static final int READ_TIMEOUT_MS = 3000;
  private static final ExecutorService EXECUTOR = Executors.newSingleThreadExecutor();

  private PushNavTargetSender() {}

  public static void sendAsync(String rawServerUrl, float raDeg, float decDeg) {
    if (rawServerUrl == null || rawServerUrl.trim().isEmpty()) {
      return;
    }
    EXECUTOR.execute(() -> {
      try {
        send(rawServerUrl, raDeg, decDeg);
      } catch (IllegalArgumentException | IOException e) {
        Log.w(TAG, "Unable to send target to PushNav", e);
      }
    });
  }

  public static void send(String rawServerUrl, float raDeg, float decDeg) throws IOException {
    validateCoordinates(raDeg, decDeg);
    String serverUrl = PushNavConnectionProbe.normalizeServerUrl(rawServerUrl);
    byte[] body = createRequestBody(raDeg, decDeg).getBytes(StandardCharsets.UTF_8);

    HttpURLConnection connection = (HttpURLConnection) new URL(serverUrl + "/api/goto/set")
        .openConnection();
    connection.setConnectTimeout(CONNECT_TIMEOUT_MS);
    connection.setReadTimeout(READ_TIMEOUT_MS);
    connection.setRequestMethod("POST");
    connection.setRequestProperty("Content-Type", "application/json; charset=utf-8");
    connection.setFixedLengthStreamingMode(body.length);
    connection.setDoOutput(true);
    connection.setUseCaches(false);

    try {
      try (OutputStream output = connection.getOutputStream()) {
        output.write(body);
      }
      int responseCode = connection.getResponseCode();
      if (responseCode != HttpURLConnection.HTTP_NO_CONTENT) {
        throw new IOException("PushNav returned HTTP " + responseCode);
      }
    } finally {
      connection.disconnect();
    }
  }

  static String createRequestBody(float raDeg, float decDeg) {
    validateCoordinates(raDeg, decDeg);
    return new JSONObject()
        .put("ra_deg", raDeg)
        .put("dec_deg", decDeg)
        .toString();
  }

  private static void validateCoordinates(float raDeg, float decDeg) {
    if (!Float.isFinite(raDeg) || raDeg < 0.0f || raDeg >= 360.0f) {
      throw new IllegalArgumentException("Right ascension must be in [0, 360)");
    }
    if (!Float.isFinite(decDeg) || decDeg < -90.0f || decDeg > 90.0f) {
      throw new IllegalArgumentException("Declination must be in [-90, 90]");
    }
  }
}

File: stardroid-v1/app/src/main/java/com/google/android/stardroid/pushnav/PushNavTargetSender.kt
package com.google.android.stardroid.pushnav

import android.util.Log
import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.Executor

object PushNavTargetSender {
  private const val TAG = "PushNavTargetSender"
  private const val CONNECT_TIMEOUT_MS = 3000
  private const val READ_TIMEOUT_MS = 3000

  @JvmStatic
  fun sendAsync(
    executor: Executor,
    rawServerUrl: String?,
    raDeg: Float,
    decDeg: Float
  ) {
    if (rawServerUrl.isNullOrBlank()) {
      return
    }
    executor.execute {
      try {
        send(rawServerUrl, raDeg, decDeg)
      } catch (exception: IllegalArgumentException) {
        Log.w(TAG, "Unable to send target to PushNav", exception)
      } catch (exception: IOException) {
        Log.w(TAG, "Unable to send target to PushNav", exception)
      }
    }
  }

  @JvmStatic
  @Throws(IOException::class)
  fun send(rawServerUrl: String?, raDeg: Float, decDeg: Float) {
    validateCoordinates(raDeg, decDeg)
    val serverUrl = PushNavConnectionProbe.normalizeServerUrl(rawServerUrl)
    val body = createRequestBody(raDeg, decDeg).toByteArray(Charsets.UTF_8)
    val connection = URL("$serverUrl/api/goto/set").openConnection() as HttpURLConnection
    connection.connectTimeout = CONNECT_TIMEOUT_MS
    connection.readTimeout = READ_TIMEOUT_MS
    connection.requestMethod = "POST"
    connection.setRequestProperty("Content-Type", "application/json; charset=utf-8")
    connection.setFixedLengthStreamingMode(body.size)
    connection.doOutput = true
    connection.useCaches = false

    try {
      connection.outputStream.use { output -> output.write(body) }
      val responseCode = connection.responseCode
      if (responseCode != HttpURLConnection.HTTP_NO_CONTENT) {
        throw IOException("PushNav returned HTTP $responseCode")
      }
    } finally {
      connection.disconnect()
    }
  }

  @JvmStatic
  fun createRequestBody(raDeg: Float, decDeg: Float): String {
    validateCoordinates(raDeg, decDeg)
    return JSONObject()
      .put("ra_deg", raDeg)
      .put("dec_deg", decDeg)
      .toString()
  }

  private fun validateCoordinates(raDeg: Float, decDeg: Float) {
    require(raDeg.isFinite() && raDeg >= 0.0f && raDeg < 360.0f) {
      "Right ascension must be in [0, 360)"
    }
    require(decDeg.isFinite() && decDeg >= -90.0f && decDeg <= 90.0f) {
      "Declination must be in [-90, 90]"
    }
  }
}
